Implement GcInfo v2

swaroop-sridhar opened the pull request:
Ref #4379

This change implements GcInfo version 2 for all platforms that use the
GcInfo library (all architectures other than X86).

Changes are:
1) Defines ReturnKind enumeration for all platforms
2) Change the GcInfo encoder library to encode the ReturnKind and ReversePInvokeFrame slot
3) Change the CM's GcInfo decoder to encode the ReturnKind and ReversePInvokeFrame slot for GCINFO_VERSION 2
4) Some corrections to GCINFO_MEASUREments
5) Changes to RYU Jit to provide the correct information to the encoder
6) Changes to the VM to use the ReturnKind information while hijacking a thread
   - If ReturnKind is available from GcInfo, new hijack routines are used
   - Otherwise, fall back to old method (for compatibility)
7) Rework and simplify the thread hijack routines by scanning HijackFrames directly for gcroots
8) Supporting code to implement the above features.

Returning Structs in multiple registers

Hijacking for StructInRegs is currently only implemented for
Unix SystemV ABI Multi-reg struct returns. However, the hijack-workers that use
ReturnKind are ready to handle other platforms (ex: ARM/ARM64 Windows)
once the corresponding HijackTripThread() assembly routines are defined.

The New feature flag: FEATURE_MULTIREG_RETURN is set for platforms where a struct value
can be returned in multiple registers [ex: Windows/Unix ARM/ARM64, Unix-AMD64]
FEATURE_UNIX_AMD64_STRUCT_PASSING is a specific kind of FEATURE_MULTIREG_RETURN
specified by SystemV ABI for AMD64

Compatibility with other JITs

- All new GCInfo generated by RYU Jit is in GcInfo version 2
- All Ngen images must be regenerated with the new GcInfo version.
- Ready-to-run images with old GcInfo will continue to work.
- Jit64/X64 uses the GcInfo library, so it generates GcInfo version 2.
  However, it doesn't (yet) provide the data to encode the correct ReturnKind
  Similar is the case for ARM32 code running on JIT32, and any other JITs
  that may be using GcInfo library but not yet modified to use the new API.
  So, compatibility is achived using RT_Unset flag.

When ReturnKind is RT_Unset, it means that the JIT did not set
the ReturnKind in the GCInfo, and therefore the VM cannot rely on it,
and must use other mechanisms (similar to GcInfo ver 1) to determine
the Return type's GC information.

Implement GC root scanning for Hijack-frames

This change implements GCScanRoots() method for Hijacke-frames
based on the ReturnKind information available from the GcInfo.
If the exact ReturnKind is not available in the GcInfo, the
thread-suspension logic will compute the ReturnKind based on
the method-signature.

As a result of this change, several hijack-helpers in the VM
are cleaned up. There's only one implementation of HijackWorker()
which handles all returnKinds, and one assembly helper for handling
all methods that return a GC value.

ARM64

Only GcIndfo v2 is reliably supported for ARM64 platform.
The changes to thread-hijack mechanism fixes #6494 for ARM64.

No measurable change in JIT throughput, performance or native-image
size from this change.

swaroop-sridhar commented:
@dotnet-bot test Windows_NT amd64 checked gcstress0xc

swaroop-sridhar commented:
@dotnet-bot test Windows_NT amd64 checked gcstress0xc

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 checked

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 checked gcstress0xc

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 checked

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 checked

swaroop-sridhar commented:
@dotnet-bot test Windows_NT gcstress0xc

swaroop-sridhar commented:
@dotnet-bot test Ubuntu gcstress0xc

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 checked pri1r2r

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 checked

swaroop-sridhar commented:
@jkotas please review.
CC: @dotnet/jit-contrib

jkotas commented on src/debug/daccess/nidump.cpp:
```diff
@@ -9440,7 +9440,7 @@ void NativeImageDumper::DumpReadyToRunMethod(PCODE pEntryPoint, PTR_RUNTIME_FUNC
         gcDump.gcPrintf = stringOutFn;
 #if !defined(_TARGET_X86_) && defined(USE_GC_INFO_DECODER)
-        GcInfoDecoder gcInfoDecoder({ curGCInfoPtr, GCINFO_VERSION }, DECODE_CODE_LENGTH, 0);
+        GcInfoDecoder gcInfoDecoder({ curGCInfoPtr, GCINFO_VERSION }, DECODE_CODE_LENGTH);
```
This is dumping method in R2R image. Should this take the version that matches the file format version? 

(It is not crucial to fix this because of it is nidump ... if it is hard to do, comment would be nice at least.)

jkotas commented on src/gcdump/gcdumpnonx86.cpp:
```diff
@@ -438,6 +439,9 @@ size_t      GCDump::DumpGCTable(PTR_CBYTE      gcInfoBlock,
 #endif
 
+    ReturnKind returnKind = hdrdecoder.GetReturnKind();
+    gcPrintf("Code size: %s\n", ReturnKindToString(returnKind));
```
Copy&paste mistake: `Code size` -> `Return kind`

jkotas commented on src/vm/threadsuspend.cpp:
```diff
@@ -7411,8 +7411,417 @@ void STDCALL OnHijackStructInRegsWorker(HijackArgs * pArgs)
+
+inline void SwapReturnRegisters(HijackArgs * pArgs)
+{
+    size_t swap = pArgs->ReturnValue[0];
```
This swapping of arguments and having number of flavors of `OnHijack*` looks pretty convoluted. 

Instead, it would be better to move the switch statement over `ReturnKind` into `HijackFrame::GcScanRoots` (new overrride of `Frame::GcScanRoots`), and have just one shared version of everything else. Something like:

```
switch (returnKind)
{
case RT_Scalar:
    // nothing to report
    break;
case RT_Object:
   (*fn)(dac_cast<PTR_PTR_Object>(&pArgs->ReturnValue[0]), sc, CHECK_APP_DOMAIN);
    break;
case RT_ByRef:
    PromoteCarefully(fn, dac_cast<PTR_PTR_Object>(&pArgs->ReturnValue[0]), sc, GC_CALL_INTERIOR | GC_CALL_CHECK_APP_DOMAIN);
    break;
...
case RT_Obj_ByRef:
   (*fn)(dac_cast<PTR_PTR_Object>(&pArgs->ReturnValue[0]), sc, CHECK_APP_DOMAIN);
   PromoteCarefully(fn, dac_cast<PTR_PTR_Object>(&pArgs->ReturnValue[1]), sc, GC_CALL_INTERIOR | GC_CALL_CHECK_APP_DOMAIN);
   break;
...
```

jkotas commented on src/vm/threadsuspend.cpp:
```diff
@@ -7411,8 +7411,417 @@ void STDCALL OnHijackStructInRegsWorker(HijackArgs * pArgs)
+
+    MethodDesc *methodDesc = codeInfo->GetMethodDesc();
+
+    if (methodDesc == nullptr) {
```
Where does this happen?

swaroop-sridhar replied:
This `GetHijackAddrFromMethodTable` method gets called:
(1) When GcInfo version is 1 (for ready to run images)
(2) When GcInfo version is 2, but ReturnKind is not set (ex: JIT64). GcInfo will report RT_Unset.
(3) Debug: Currently used to verify that `GetHijackAddrFromMethodTable` and `GetHijackAddrFromGcInfo` return the same answer.

jkotas replied:
I meant - when is `methodDesc == nullptr`?

swaroop-sridhar replied:
I'm not sure how it can be null. 
I wrote this case to mimic the existing behavior in [` Thread::HandledJITCase`](https://github.com/dotnet/coreclr/blob/master/src/vm/threadsuspend.cpp#L7868-L7909) which I refactored out to this function.

jkotas replied:
I would change this to assert.

swaroop-sridhar commented:
@jkotas, I've addressed all the above comments in this commit:
https://github.com/dotnet/coreclr/pull/6522/commits/24135e1902bc8df8a8c6cb647e88d01d1ce05a5b 
Please review.

swaroop-sridhar commented:
@dotnet-bot test Windows_NT amd64 checked gcstress0xc

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 checked pri1r2r

swaroop-sridhar commented:
@dotnet-bot test Ubuntu gcstress0xc
@dotnet-bot test Windows_NT arm64 checked
@dotnet-bot test Windows_NT gcstress0xc

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 Checked

swaroop-sridhar commented:
@dotnet-bot test  Windows_NT x64 Debug Build and Test

swaroop-sridhar commented:
@dotnet-bot  test Windows_NT x64 Checked Build and Test (Jit - GCStress=0xC)

swaroop-sridhar commented:
@dotnet-bot test Ubuntu gcstress0xc

jkotas commented:
FYI @janvorli It is changing how GC reporting of return values during hijacking works

janvorli commented on src/vm/amd64/unixasmhelpers.S:
```diff
@@ -304,7 +232,7 @@ NESTED_ENTRY OnHijackInteriorPointerTripThread, _TEXT, NoHandler
+    call                C_FUNC(OnHijackWorker)
 
     free_stack          0x8
     pop_register        rax
```
How does the GC update the RDX if it needs to? The original OnHijackStructInRegsTripThread was pushing RAX and RDX.

swaroop-sridhar replied:
`RDX` and `RAX` are pushed on to the stack by this `OnHijackGCTripThread()` helper.
So, `OnHijackWorker()` and eventually `HijackFrame::GCScanRoots()` will recieve them as `m_Args->ReturnValue[2]`. 
The second register `RDX` will be updated by this case:
https://github.com/dotnet/coreclr/pull/6522/files#diff-bffe9cf484a03aa5d9d19d6842a144b1R1175

jkotas replied:
You are right - `RDX` and `RAX` are pushed here.

There may be a different problem: Is it possible to have a stack that contains both GC reference and floating point value (xmm0/xmm1)? OnHijackGCTripThread is not preserving floating point return values.

Do we really need both Scalar and GC asm helpes? One universal one should be enough. This is not performance critical path - having one universal helper and smaller/simpler code is better. The only case that needs non-standard hijack helper is x86 floating point return value to deal with floating point stack. Everything else can use the universal one.

janvorli replied:
Ah, I am sorry, the diff has hidden the RDX manipulation and I have missed that.
As for the possibility of GC reference and floating point in the return value, yes, it is possible. It is valid to return a struct with one reference and one floating point in registers.

swaroop-sridhar replied:
@jkotas: I've implemented your recommendation in this commit: https://github.com/dotnet/coreclr/pull/6522/commits/f10be428c72a2d22fd59443797cdde59102dc9ea

swaroop-sridhar commented:
@dotnet-bot test Windows_NT gcstress0xc
@dotnet-bot test Ubuntu gcstress0xc
@dotnet-bot test Windows_NT arm64 Checked
@dotnet-bot test Windows_NT arm64 checked pri1r2r
@dotnet-bot test Windows_NT arm64 Release
@dotnet-bot test Windows_NT arm64 checked gcstress0xc

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 Checked

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 checked gcstress0xc

swaroop-sridhar commented:
@dotnet-bot test Windows_NT gcstress0xc
@dotnet-bot test Ubuntu gcstress0xc
@dotnet-bot test Windows_NT arm64 Checked
@dotnet-bot test Windows_NT arm64 checked pri1r2r
@dotnet-bot test Windows_NT arm64 Release
@dotnet-bot test Windows_NT arm64 checked gcstress0xc

jkotas commented:
👍

janvorli commented:
LGTM

swaroop-sridhar commented:
@dotnet-bot test Ubuntu gcstress0xc

swaroop-sridhar commented:
@dotnet-bot test Windows_NT gcstress0xc
@dotnet-bot test Ubuntu gcstress0xc
@dotnet-bot test Windows_NT arm64 Checked
@dotnet-bot test Windows_NT arm64 checked pri1r2r
@dotnet-bot test Windows_NT arm64 Release
@dotnet-bot test Windows_NT arm64 checked gcstress0xc

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 Checked
@dotnet-bot test Windows_NT arm64 checked gcstress0xc

swaroop-sridhar commented:
@dotnet-bot  test OSX x64 Checked Build and Test

swaroop-sridhar commented:
@dotnet-bot  test Ubuntu x64 Checked Build and Test

swaroop-sridhar commented:
@dotnet-bot  test Linux ARM Emulator Cross Debug Build

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 Checked

briansull commented on src/vm/threadsuspend.cpp:
```diff
@@ -7180,238 +7178,142 @@ void STDCALL OnHijackObjectWorker(HijackArgs * pArgs)
+            }
+        }
+        ReturnKind structReturnKind = GetStructReturnKind(regKinds[0], regKinds[1]);
+        return structReturnKind;
```
Is there equivalent functionality for ARM64 someplace here?

swaroop-sridhar replied:
This code only exists to support GcInfo v1 images, which are not supported for ARM64.
I'll add a comment about it.
ARM64 struct-in-regs scanning happens in HijackFrame::GcScanRoots()

briansull commented on src/vm/threads.h:
```diff
@@ -1016,13 +1012,20 @@ typedef DWORD (*AppropriateWaitFunc) (void *args, DWORD timeout, DWORD option);
+// Unix SystemV ABI Multi-reg struct returns. However, the hijack-workers that use 
+// ReturnKind are ready to handle other platforms (ex: ARM/ARM64 Windows) 
+// once the corresponding HijackTripThread() assembly routines are defined.
+
```
OK I see that we are still waiting on ARM64 functionality.

swaroop-sridhar replied:
I'll remove the comment -- its actually now stale

briansull commented:
LGTM

Is there a issue open for the remaining Arm64 work?

swaroop-sridhar commented:
You mean the ARM64 work for struct-in-regs return, right? 
There's and issue for it #6494. and it'll be fixed by this change.

swaroop-sridhar commented:
Windows NT GCStress 0xC:
The two failures look like they are pre-existing:
1) Random failure writing to protected memory in Console.WriteLine()
2) Failure in GetGenerationWR2.exe
 [Here](http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/job/jitstress/job/x64_checked_windows_nt_gcstress0xc/7/consoleText) is the last week's run for comparison.

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 Checked

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 Checked

swaroop-sridhar commented:
@dotnet-bot  test Linux ARM Emulator Cross Debug Build

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 Checked

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 Checked

swaroop-sridhar commented:
@dotnet-bot  Details Linux ARM Emulator Cross Debug Build

swaroop-sridhar commented:
@dotnet-bot test Windows_NT arm64 checked
